libyul/ObjectCache.cpp

```diff
@@ -0,0 +1,143 @@
+	case Language::Yul:
+		return EVMDialectTyped::instance(_version);
+	}
+	util::unreachable();
+}
+
+void ObjectCache::optimize(Object& _object, Context const& _context, bool _isCreation)
+{
+	yulAssert(_object.code);
+	yulAssert(_object.analysisInfo); // TMP: Do I need this?
+	yulAssert(_object.debugData);
+
+	for (auto& subNode: _object.subObjects)
+		if (auto subObject = dynamic_cast<Object*>(subNode.get()))
+		{
+			bool isCreation = !boost::ends_with(subObject->name, "_deployed");
+			optimize(
+				*subObject,
+				_context,
+				isCreation
+			);
+		}
+
+	Dialect const& dialect = languageToDialect(_context.language, _context.evmVersion);
+	std::unique_ptr<GasMeter> meter;
+	if (EVMDialect const* evmDialect = dynamic_cast<EVMDialect const*>(&dialect))
+		meter = std::make_unique<GasMeter>(*evmDialect, _isCreation, _context.expectedExecutionsPerDeployment);
+
+	h256 cacheKey = calculateCacheKey(*_object.code, *_object.debugData, _context, _isCreation);
+	if (m_cachedObjects.count(cacheKey) != 0)
+	{
+		retrieveObject(cacheKey, _object);
+		return;
+	}
+
+	OptimiserSuite::run(
+		dialect,
+		meter.get(),
+		_object,
+		// Defaults are the minimum necessary to avoid running into "Stack too deep" constantly.
+		_context.optimizeStackAllocation,
+		_context.yulOptimiserSteps,
+		_context.yulOptimiserCleanupSteps,
+		_isCreation ? std::nullopt : std::make_optional(_context.expectedExecutionsPerDeployment),
+		{}
+	);
+
+	storeObject(cacheKey, _object);
+}
+
+void ObjectCache::storeObject(util::h256 _cacheKey, Object& _object)
+{
+	m_cachedObjects[_cacheKey] = CachedObject{
+		std::make_shared<Block>(ASTCopier{}.translate(*_object.code)),
+		_object.analysisInfo,
+		_object.debugData,
+	};
+}
+
+void ObjectCache::retrieveObject(util::h256 _cacheKey, Object& _object) const
+{
+	solAssert(m_cachedObjects.count(_cacheKey) != 0);
+	_object.code = std::make_shared<Block>(ASTCopier{}.translate(*m_cachedObjects.at(_cacheKey).ast));
+	_object.analysisInfo = m_cachedObjects.at(_cacheKey).analysisInfo;
+	_object.debugData = m_cachedObjects.at(_cacheKey).debugData;
+	solAssert(_object.code);
+	solAssert(_object.analysisInfo);
+	solAssert(_object.debugData);
+}
+
+h256 ObjectCache::calculateCacheKey(
+	Block const& _ast,
+	ObjectDebugData const& _debugData,
+	Context const& _context,
+	bool _isCreation
+)
+{
+	// TMP: Is the code indented deterministically?
+	AsmPrinter asmPrinter(nullptr, _debugData.sourceNames, DebugInfoSelection::All());
+
+	bytes rawKey;
+	rawKey += keccak256(asmPrinter(_ast)).asBytes();
+	rawKey += h256(u256(_context.language)).asBytes();
+	rawKey += h256(u256(_context.optimizeStackAllocation ? 0 : 1)).asBytes();
+	rawKey += h256(u256(_context.expectedExecutionsPerDeployment)).asBytes();
+	rawKey += h256(u256(_isCreation ? 0 : 1)).asBytes();
+	rawKey += keccak256(_context.evmVersion.name()).asBytes();
+	rawKey += keccak256(_context.yulOptimiserSteps).asBytes();
+	rawKey += keccak256(_context.yulOptimiserCleanupSteps).asBytes();
```

Comment:
Weird, I'm sure I had another comment about whether the cache should be the one performing the optimization, but apparently it got lost. In any case: that looked odd at first, but it does make sense, since it guarantees that the cached object *does* correspond to the right settings and *only* the right settings - and then this hashing scheme makes sense, so I'm resolving the comment.

Reply:
I initially did it like you suggested in your first comment, i.e. only hashing the input AST and resetting the cache in `CompilerStack`, but I did not like how brittle it felt. You need to be more careful about when you reset - that's needed not only in `reset()` but also in a setter of every parameter that might affect the result. I think that's too disconnected from the actual optimization logic.

The current design does more things but makes usage simpler and more fool-proof. I think that's important especially if you're thinking about extending the mechanism to two-step compilation, where the cache would be exposed to the outside. In the current form it's actually pretty general and technically there could exist a single instance of the cache for the whole compiler or even across multiple compiler runs.

Reply:
> whether the cache should be the one performing the optimization

Actually, now I see it more like an extension of the optimizer `Suite` than just cache. I considered making this a part of `Suite` but for now decided not to. I'm still going to rename it to something like `CachingOptimizer` (or maybe I'll find a better name) since now it's a part of the optimizer logic.

Reply:
Yep, as I said, it was just an initial reaction, but your reasoning is sound, I agree!